File: qms-core-mfe/src/app/nonconformance/services/nonconformance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Page } from '../../shared/core-lookup.service';

export interface Nonconformance {
  id: string;
  ncNumber: string;
  title: string;
  description: string;
  ncType: string;
  classification: string | null;
  status: string;
  priority: string;
  productName: string | null;
  batchNumber: string | null;
  quantityAffected: string | null;
  stageDetected: string | null;
  dispositionDecision: string | null;
  dispositionJustification: string | null;
  dispositionDate: string | null;
  holdStatus: string;
  holdLocation: string | null;
  holdInitiatedDate: string | null;
  holdReleasedDate: string | null;
  capaRequired: boolean;
  owner: { id: string; displayName: string } | null;
  department: { id: string; name: string } | null;
  plantSite: { id: string; name: string } | null;
  createdAt: string;
  updatedAt: string;
}

@Injectable({ providedIn: 'root' })
export class NonconformanceService {
  private readonly baseUrl = 'http://localhost:8082/api/v1/nonconformances';

  constructor(private http: HttpClient) {}

  list(params?: Record<string, string>): Observable<Page<Nonconformance>> {
    return this.http.get<Page<Nonconformance>>(this.baseUrl, { params: this.toParams(params) });
  }

  getById(id: string): Observable<Nonconformance> {
    return this.http.get<Nonconformance>(`${this.baseUrl}/${id}`);
  }

  create(request: Record<string, unknown>): Observable<Nonconformance> {
    return this.http.post<Nonconformance>(this.baseUrl, request);
  }

  update(id: string, request: Record<string, unknown>): Observable<Nonconformance> {
    return this.http.put<Nonconformance>(`${this.baseUrl}/${id}`, request);
  }

  transitionStatus(id: string, status: string): Observable<Nonconformance> {
    return this.http.patch<Nonconformance>(`${this.baseUrl}/${id}/status`, { status });
  }

  submitDisposition(id: string, request: Record<string, unknown>): Observable<Nonconformance> {
    return this.http.patch<Nonconformance>(`${this.baseUrl}/${id}/disposition`, request);
  }

  toggleHold(id: string, request: Record<string, unknown>): Observable<Nonconformance> {
    return this.http.patch<Nonconformance>(`${this.baseUrl}/${id}/hold`, request);
  }

  getDashboard(): Observable<Record<string, unknown>> {
    return this.http.get<Record<string, unknown>>(`${this.baseUrl}/dashboard`);
  }

  private toParams(params?: Record<string, string>): HttpParams {
    let p = new HttpParams();
    if (params) Object.entries(params).forEach(([k, v]) => { if (v) p = p.set(k, v); });
    return p;
  }
}

File: backend/src/main/java/com/qmspharma/service/NonconformanceService.java
package com.qmspharma.service;

import com.qmspharma.model.entity.*;
import com.qmspharma.repository.*;
import com.qmspharma.security.CurrentUserProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.*;

@Service
@RequiredArgsConstructor
public class NonconformanceService {

    private static final String RECORD_TYPE = "NONCONFORMANCE";

    private final NonconformanceRepository ncRepository;
    private final PlantSiteRepository plantSiteRepository;
    private final DepartmentRepository departmentRepository;
    private final UserRepository userRepository;
    private final SequenceGeneratorService sequenceGenerator;
    private final AuditTrailService auditTrailService;
    private final WorkflowService workflowService;
    private final CurrentUserProvider currentUserProvider;

    @Transactional(readOnly = true)
    public Page<Nonconformance> list(String status, String ncType, String holdStatus,
                                      UUID plantSiteId, String search, Pageable pageable) {
        Specification<Nonconformance> spec = Specification.where(null);
        if (status != null) spec = spec.and((r, q, cb) -> cb.equal(r.get("status"), status));
        if (ncType != null) spec = spec.and((r, q, cb) -> cb.equal(r.get("ncType"), ncType));
        if (holdStatus != null) spec = spec.and((r, q, cb) -> cb.equal(r.get("holdStatus"), holdStatus));
        if (plantSiteId != null) spec = spec.and((r, q, cb) -> cb.equal(r.get("plantSite").get("id"), plantSiteId));
        if (search != null && !search.isBlank()) {
            String like = "%" + search.toLowerCase() + "%";
            spec = spec.and((r, q, cb) -> cb.or(
                    cb.like(cb.lower(r.get("title")), like),
                    cb.like(cb.lower(r.get("ncNumber")), like)
            ));
        }
        return ncRepository.findAll(spec, pageable);
    }

    @Transactional(readOnly = true)
    public Nonconformance getById(UUID id) {
        return ncRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Nonconformance not found: " + id));
    }

    @Transactional
    public Nonconformance create(Map<String, Object> request) {
        User currentUser = currentUserProvider.getCurrentUser();
        Nonconformance nc = new Nonconformance();
        nc.setNcNumber(sequenceGenerator.generateNumber("NONCONFORMANCE"));
        nc.setTitle((String) request.get("title"));
        nc.setDescription((String) request.get("description"));
        nc.setNcType((String) request.get("ncType"));
        nc.setPriority((String) request.getOrDefault("priority", "MEDIUM"));
        nc.setOwner(userRepository.getReferenceById(UUID.fromString((String) request.get("ownerId"))));
        nc.setDepartment(departmentRepository.getReferenceById(UUID.fromString((String) request.get("departmentId"))));
        nc.setPlantSite(plantSiteRepository.getReferenceById(UUID.fromString((String) request.get("plantSiteId"))));
        if (request.containsKey("productName")) nc.setProductName((String) request.get("productName"));
        if (request.containsKey("batchNumber")) nc.setBatchNumber((String) request.get("batchNumber"));
        if (request.containsKey("stageDetected")) nc.setStageDetected((String) request.get("stageDetected"));
        nc.setCreatedBy(currentUser);
        nc.setUpdatedBy(currentUser);

        Nonconformance saved = ncRepository.save(nc);
        workflowService.startProcess("nonconformanceProcess", saved.getId().toString(), Map.of());
        return saved;
    }

    @Transactional
    public Nonconformance update(UUID id, Map<String, Object> request) {
        Nonconformance nc = getById(id);
        if (request.containsKey("title")) nc.setTitle((String) request.get("title"));
        if (request.containsKey("classification")) nc.setClassification((String) request.get("classification"));
        if (request.containsKey("quantityAffected")) nc.setQuantityAffected((String) request.get("quantityAffected"));
        nc.setUpdatedBy(currentUserProvider.getCurrentUser());
        return ncRepository.save(nc);
    }

    @Transactional
    public Nonconformance transitionStatus(UUID id, String newStatus) {
        Nonconformance nc = getById(id);
        String old = nc.getStatus();
        nc.setStatus(newStatus);
        if ("CLOSED".equals(newStatus)) nc.setClosedDate(Instant.now());
        nc.setUpdatedBy(currentUserProvider.getCurrentUser());
        auditTrailService.logAction(RECORD_TYPE, nc.getId(), nc.getNcNumber(),
                "STATUS_CHANGE", "status", old, newStatus, null);
        return ncRepository.save(nc);
    }

    @Transactional
    public Nonconformance submitDisposition(UUID id, Map<String, Object> request) {
        Nonconformance nc = getById(id);
        nc.setDispositionDecision((String) request.get("dispositionDecision"));
        nc.setDispositionJustification((String) request.get("justification"));
        nc.setDispositionDate(Instant.now());
        nc.setDispositionApprovedBy(currentUserProvider.getCurrentUser());
        nc.setStatus("DISPOSITION_APPROVED");
        nc.setUpdatedBy(currentUserProvider.getCurrentUser());
        return ncRepository.save(nc);
    }

    @Transactional
    public Nonconformance toggleHold(UUID id, Map<String, Object> request) {
        Nonconformance nc = getById(id);
        String action = (String) request.get("action");
        if ("HOLD".equals(action)) {
            nc.setHoldStatus("HOLD");
            nc.setHoldLocation((String) request.get("holdLocation"));
            nc.setHoldInitiatedDate(Instant.now());
        } else {
            nc.setHoldStatus("RELEASED");
            nc.setHoldReleasedDate(Instant.now());
            nc.setHoldReleasedBy(currentUserProvider.getCurrentUser());
        }
        nc.setUpdatedBy(currentUserProvider.getCurrentUser());
        return ncRepository.save(nc);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getDashboardMetrics() {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("totalNonconformances", ncRepository.count());
        m.put("onHold", ncRepository.countOnHold());
        m.put("byStatus", ncRepository.countByStatusGrouped());
        m.put("byType", ncRepository.countByNcType());
        m.put("byDisposition", ncRepository.countByDisposition());
        return m;
    }
}
